File: FactoryMethodPattern/python/PizzaStore.py
import abc
from enum import Enum

class PizzaType(Enum):
    CHEESE_PIZZA = "cheese flavor pizza"
    PEPPERONI_PIZZA = "pepperoni flavor pizza"
    VEGGIE_PIZZA = "veggie flavor pizza"


class Pizza:
    def __init__(self, radius:int) -> None:
        self._radius = radius

    @property
    @abc.abstractmethod
    def price(self) -> int:
        return NotImplemented
    
    @property
    def radius(self) -> int:
        return self._radius


class CheesePizza(Pizza):
    def __init__(self, radius) -> None:
        super().__init__(radius)

    @property
    def price(self):
        return (self._radius**2) * 10


class PepperoniPizza(Pizza):
    def __init__(self, radius) -> None:
        super().__init__(radius)

    @property
    def price(self):
        return (self._radius**2) * 20


class VeggiePizza(Pizza):
    def __init__(self, radius) -> None:
        super().__init__(radius)

    @property
    def price(self):
        return (self._radius**2) * 15


class PizzaStore:
    @abc.abstractmethod
    def createPizza(self, type:PizzaType, radius:int) -> Pizza:
        return NotImplemented
    
    def orderPizza(self, pizzaType:PizzaType, radius:int) -> Pizza:
        pizza = self.createPizza(pizzaType, radius)
        if pizza:
            print(f"The {pizzaType.value} is provided. The radius of pizza is {radius} and the price of pizza is {pizza.price}")
        else:
            print(f"The {pizzaType.value} is not provided !!!")

        return pizza


class PizzaStoreA(PizzaStore):
    def createPizza(self, pizzaType:PizzaType, radius:int) -> Pizza:
        if pizzaType == PizzaType.CHEESE_PIZZA:
            return CheesePizza(radius)
        elif pizzaType == PizzaType.PEPPERONI_PIZZA:
            return PepperoniPizza(radius)
    

class PizzaStoreB(PizzaStore):
    def createPizza(self, pizzaType:PizzaType, radius:int) -> Pizza:
        if pizzaType == PizzaType.CHEESE_PIZZA:
            return CheesePizza(radius)
        elif pizzaType == PizzaType.VEGGIE_PIZZA:
            return VeggiePizza(radius)
        

File: FactoryMethodPattern/cpp/src/pizza_store.cpp
#include <iostream>
#include "header/pizza_store.h"

using namespace std;

string pizzaTypeToString(PizzaType type){
    if(type == CHEESE_PIZZA)
        return "cheese flavor pizza";
    else if(type == PEPPERONI_PIZZA)
        return "pepperoni flavor pizza";
    else if(type == VEGGIE_PIZZA)
        return "veggie flavor pizza";

    return "";
}

int Pizza::getRadius(){
    return radius;
}

CheesePizza::CheesePizza(int radius){
    this->radius = radius;
}

int CheesePizza::getPrice(){
    return radius * radius * 10;
}

PepperoniPizza::PepperoniPizza(int radius){
    this->radius = radius;
}

int PepperoniPizza::getPrice(){
    return radius * radius * 20;
}

VeggiePizza::VeggiePizza(int radius){
    this->radius = radius;
}

int VeggiePizza::getPrice(){
    return radius * radius * 15;
}

shared_ptr<Pizza> PizzaStore::orderPizza(PizzaType type, int radius){
    shared_ptr<Pizza> pizza = createPizza(type, radius);
    if(pizza == nullptr){
        cout << "The " << pizzaTypeToString(type) << " is not provided !!!" << endl;
    }else{
        cout << "The " << pizzaTypeToString(type) << " is provided. The radius of pizza is " << pizza->getRadius() << " and the price of pizza is " << pizza->getPrice() << endl;
    }

    return pizza;
}

shared_ptr<Pizza> PizzaStoreA::createPizza(PizzaType type, int radius){
    if(type == CHEESE_PIZZA)
        return make_shared<CheesePizza>(radius);
    else if(type == PEPPERONI_PIZZA)
        return make_shared<PepperoniPizza>(radius);

    return shared_ptr<Pizza>();
}

shared_ptr<Pizza> PizzaStoreB::createPizza(PizzaType type, int radius){
    if(type == CHEESE_PIZZA)
        return make_shared<CheesePizza>(radius);
    else if(type == VEGGIE_PIZZA)
        return make_shared<VeggiePizza>(radius);

    return shared_ptr<Pizza>();
}